### DISCUSSION TITLE
How do I include additional libs (e.g. libserialport) in the toolchain?

### DISCUSSION THREAD

**Discussion by toomanybrians** [abhiTronix/raspberry-pi-cross-compilers#111] [Q&A] [answered]
Hello! First, thank you for this awesome project and wonderful instructions. I am new to cross compiling, so hopefully this question is easy.

I have an app I'm trying to compile for Raspberry Pi via GitHub actions using an Ubuntu runner. The complexity for me is our app needs `libserialport`. I added libserialport-dev to the apt-get and updated the include flag in the compiler step, so I'm able to successfully compile.

However I get an error on linking that -lserialport is not found. (Actually we use curses too so I also get an error that -lcurses is not found.)

If I add `ld -lcurses --verbose` and `ld -lserialport --verbose` to our build script, I can see all the locations where these libs can be found, but of course it finds both the x86_64 versions which are valid for the host and, for curses, the arm version in `cross-pi-gcc-10.3.0-64/lib`.

I assume I need to link against the arm versions, which I attempted to do by specifying that path as an `-L` flag. But that doesn't work. Also for libserialport which is not part of the toolchain by default, how do I get the arm binary of that? Should I build from source using this toolchain first? Or find the version that apt-get would download if I ran that from the RPi?

Here is my workflow.yml file (well, the relevant portions), so you can see what I'm doing.

```
sudo apt update && sudo apt dist-upgrade
sudo apt-get install build-essential gawk gcc g++ gfortran git texinfo bison libncurses-dev libserialport-dev
curl -L -O -s https://sourceforge.net/projects/raspberry-pi-cross-compilers/files/Bonus%20Raspberry%20Pi%20GCC%2064-Bit%20Toolchains/Raspberry%20Pi%20GCC%2064-Bit%20Cross-Compiler%20Toolchains/Bullseye/GCC%2010.3.0/cross-gcc-10.3.0-pi_64.tar.gz/download
tar xf download
export PATH=$(pwd)/cross-pi-gcc-10.3.0-64/bin:$PATH
export LD_LIBRARY_PATH=$(pwd)/cross-pi-gcc-10.3.0-64/lib:$LD_LIBRARY_PATH
ld -lcurses --verbose
ld -lserialport --verbose
aarch64-linux-gnu-gcc -O3 -Wall -I /usr/include -L /lib/x86_64-linux-gnu -L $(pwd)/cross-pi-gcc-10.3.0-64/lib -o my_app my_app.c -lcurses -lserialport
``` 

(The ld commands are just for debugging trying to figure out what's what in the logs)

I'd appreciate any insight as to whether I'm taking the right approach here, or if there's anything else I need to understand.

Thanks in advance for any help or guidance!
Brian

**Comment by abhiTronix** [comment, marked as answer]
> export PATH=$(pwd)/cross-pi-gcc-10.3.0-64/bin:$PATH
> export LD_LIBRARY_PATH=$(pwd)/cross-pi-gcc-10.3.0-64/lib:$LD_LIBRARY_PATH
> ld -lcurses --verbose
> ld -lserialport --verbose
> aarch64-linux-gnu-gcc -O3 -Wall -I /usr/include -L /lib/x86_64-linux-gnu -L $(pwd)/cross-pi-gcc-10.3.0-64/lib -o my_app my_app.c -lcurses -lserialport

@toomanybrians You cannot build files like this if you're cross-compiling for different target OS than Host OS. You need to download/rsync sysroot from the target OS, and then cross-compile the binaries (in your case `libserialport`). See this hello project for reference: https://github.com/abhiTronix/raspberry-pi-cross-compilers/wiki/Cross-Compiler-CMake-Usage-Guide-with-rsynced-Raspberry-Pi-32-bit-OS#cross-compiler-cmake-usage-guide-with-rsynced-raspberry-pi-32-bit-os

**Reply by toomanybrians** [reply]
Thank for this guidance! I am so new I didn’t even realize this tutorial applied to me.

I see there are 32-bit and 64-bit versions available, awesome!

One final question, if I want to do these builds on GitHub-hosted cloud runners, do you have any hints on the best approach to take? (Just point me in the right direction and I will google and figure it out.) But is it using docker versus native on the ubuntu runner? And do I need to find an RPi image to extract the sysroot components at build time, or?

Again, thank you!
Brian

**Reply by abhiTronix** [reply]
> One final question, if I want to do these builds on GitHub-hosted cloud runners, do you have any hints on the best approach to take? (Just point me in the right direction and I will google and figure it out.)

For cloud runner, use a docker with sysroot built from image(see wiki link this comment), you can find many example/template docker if you put this repo links in github search: https://github.com/search?l=Dockerfile&q=https%3A%2F%2Fsourceforge.net%2Fprojects%2Fraspberry-pi-cross-compilers&type=Code

> But is it using docker versus native on the ubuntu runner?

For native only. For docker see my other wiki link in this comment.

> And do I need to find an RPi image to extract the sysroot components at build time, or?

For building sysroot for image, see this doc instead: https://github.com/abhiTronix/raspberry-pi-cross-compilers/wiki/Cross-Compiler-CMake-Usage-Guide-with-Raspbian-32-bit-Image#cross-compiler-cmake-usage-guide-with-raspbian-32-bit-image

It contains a bash script that will automatically build sysroot from downloaded image.

**Reply by abhiTronix** [reply]
But first try building your project natively on a linux machine, if it is successful move to docker. Then once docker file is working locally too, deploy it to server. That's the best approach.

**Reply by toomanybrians** [reply]
Thank you so much for this guidance. You are awesome!
